docs/enhancements/dual-stack-vips.md

```diff
@@ -0,0 +1,142 @@
+
+### Implementation Details
+
+Our change can be simply described as "making a singular field plural". For this, we will follow the [Kubernetes API change guide](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#making-a-singular-field-plural).
+
+#### Phase 1
+
+`cluster-create-params`, `v2-cluster-update-params` and `cluster` will get new fields
+
+* `api_vips` of type `array`
+* `ingress_vips` of type `array`
+
+Respective AgentClusterInstall CR will also get those new fields.
+
+The database schema will be updated by creating new tables `api_vips` and `ingress_vips` with the following columns
+
+* `cluster_id`
+* `ip`
+
+Given that all the newly introduced fields are optional, there is no need to release a new version of the API.
+
+#### Phase 2
+
+A set of functions for cluster creation will provide a compatiblity between already existing fields (`api_vip` and `ingress_vip`) and new ones (`api_vips` and `ingress_vips`). The behaviour will be as follows
+
+* If only `api_vip` is provided, the `api_vips` will be automatically populated by a simple conversion from `string` to `[]string`. This guarantees no change in behaviour for users not aware of introduction of the new field.
+
+* If both `api_vip` and `api_vips` are provided, the value of `api_vips[0]` must match the value of `api_vip`.
+
+* If only `api_vips` are provided, the error will be raised. This is the case when plural field is specified and the singular is not. The rationale is that in an update it's impossible to tell the difference between an old client clearing the singular field via patch and a new client setting the plural field. The update semantics must not differ from create, therefore behaviour as described.
+
+For cluster update the following logic will apply
+
+* If `api_vip` is cleared and `api_vips` are not provided, we will clear both fields. This is a scenario when an old client (not aware of plural field) is used.
+
+* If `api_vips` is cleared and `api_vip` is not provided, the change will not apply. This is a scenario when an old client is used and it can't send fields it doesn't know about.
+
+* If `api_vip` is changed (but not cleared), `api_vips` will be forcefully set to the value of `api_vip` as a one-element list. This is a scenario when an old client is used to change the field it knows about.
+
+All the points described above will behave in an equivalent way for `ingress_vip` and `ingress_vips`.
+
+#### Phase 3
+
+After providing a required period of backwards-compatibility the `api_vip` and `ingress_vip` fields will be deprecated and removed. The value in the database will be nulled. Removal of the column in the database will follow.
+
+The new version of API will be released as we are removing an existing field.
+
+### UI Impact
+
+UI changes will be required in order to
+
+* display details of dual-stack clusters with dual-stack VIPs correctly
```

Comment:
On that regard, the proposed changes are going to be backward-compatible. Old clients keep working as usual with the singular fields until we announce they'll be deprecated and new clients will use both fields single and plural whenever they need to issue update or create calls, that's the TL;DR of this proposal.  
BTW, keep in mind that the web UI, given it's a web client, is going to transition to the new API faster, users will only need to refresh the browser in order to get the latest client version, once it is implemented)

Reply:
Because of all the magic dance we are doing in Phase 2, every tool we have will be backwards-compatible. If we had the possibility to skip that, we would move from Phase 1 directly to Phase 3